_posts/2019-01-14-Infrastructure-as-code-with-terraform-and-aws-serverless.md

```diff
@@ -0,0 +1,497 @@
+Without going to deep into it the aws permission model works as follows:
+* you give a resource a role
+* to this role you add permissions
+* these permissions give it access to other resources:
+    * either permission for triggering an other resource (eg. Lambda Function forwards logs to CloudWatch)
+    * or permission for being triggered by an other resource (eg. Lambda Function may be triggered by ApiGateway)
+
+```hcl-terraform
+# ROLES
+# IAM role which dictates what other AWS services the Lambda function
+# may access.
+resource "aws_iam_role" "lambda-iam-role" {
+  name = "codingtips_lambda_role"
+
+  assume_role_policy = <<EOF
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Action": "sts:AssumeRole",
+      "Principal": {
+        "Service": "lambda.amazonaws.com"
+      },
+      "Effect": "Allow",
+      "Sid": ""
+    }
+  ]
+}
+EOF
+}
+
+# POLICIES
+resource "aws_iam_role_policy" "dynamodb-lambda-policy"{
+  name = "dynamodb_lambda_policy"
+  role = "${aws_iam_role.lambda-iam-role.id}"
+  policy = <<EOF
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Action": [
+        "dynamodb:*"
+      ],
+      "Resource": "${aws_dynamodb_table.codingtips-dynamodb-table.arn}"
+    }
+  ]
+}
+EOF
+}
+
+resource "aws_iam_role_policy" "cloudwatch-lambda-policy"{
+  name = "cloudwatch-lambda-policy"
+  role = "${aws_iam_role.lambda-iam-role.id}"
+  policy = "${data.aws_iam_policy_document.api_gateway_logs_policy_document.json}"
+}
+
+data "aws_iam_policy_document" "api_gateway_logs_policy_document" {
+  statement {
+    actions = [
+      "logs:CreateLogStream",
+      "logs:CreateLogGroup",
+      "logs:PutLogEvents"
+    ],
+    resources = [
+      "arn:aws:logs:*:*:*"
+    ]
+  }
+}
+
+```
+
+In the example above the first resource that is defined is an `aws_iam_role`.
+This is the role that we will later give to our Lambda Functions.
+
+Then, we create `aws_iam_role_policy` resources to attach the role.
+The first `aws_iam_role_policy` is giving this role permission to invoke any action on the specified DynamoDB resource.
+The second role_policy allows a resource with this role to send logs to CloudWatch.
+
+A couple of things to notice:
+* The `aws_iam_role` and the `aws_iam_role_policy` are connected by the `role` argument of the role_policy resource
+* In the `statement` attribute of the `aws_iam_role_policy` we grant (Effect attr.) permission to do some actions (Action attr.) on a certain Resource (Resource attr.)
+* A resource is referenced by its *arn* or *Amazon Resource Name* which uniquely identifies this resource on aws
+* There are two ways to specify an `aws_iam_role_policy`: 
+    * the first using the *untill EOF* syntax
+    * the second using a separate terraform `data` element
+* The dynamodb-lambda-policy allows all actions on the specified dynamodb resource because under the action attribute is states `dynamodb:*`
```

Comment:
```suggestion
* The dynamodb-lambda-policy allows all actions on the specified dynamodb resource because under the action attribute it states `dynamodb:*`
```